fix(saml): send the authenticated NameID (+ format/qualifiers/SessionIndex) in the SLO LogoutRequest

Drago37 opened the pull request:
## Description

SAML/OpenID logout was broken: the IdP was never properly notified on user logout (the local session was invalidated before the provider logout could read the SAML identifiers it needs).

Concretely, the SAML `LogoutRequest` carried `NameID = <IdP metadata URL>` with `Format="...:entity"` and **no `SessionIndex`**, instead of the authenticated user's `NameID`. A strict IdP (e.g. OneLogin) rejects it with `NameID '...metadata/' does not match prior auth request`. (Keycloak is tolerant and hides the bug.)

### Changed

* SLO initiation moved out of the `WriteSessionRepository` (no business logic in the repository) into the `LogoutSession` use case
* SAML provider logout runs **before** local invalidation (reads `$_SESSION['saml']`); OpenID logout runs **after**
* `LogoutFromIdp` (`/saml/sls`) now only processes the IdP callback, instead of re-initiating a logout
* `OneLogin\Saml2\Auth` instantiation extracted behind an injectable `SamlAuthFactory` (testability)
* error handling and logging improved

### Fixed

* logout doesn't work: added the missing `samlNameIdFormat`, `nameIdNameQualifier`, `samlNameIdSPNameQualifier` to the LogoutRequest
* session invalidation ordering (was wiping the session before reading the SAML identifiers)
* `handleCallbackLogoutResponse` now invalidates the local session on the IdP callback (`keepLocalSession: false`)

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [x] 24.10.x
- [x] develop

<h2> How this pull request can be tested ? </h2>

> ⚠️ Keycloak (the default dev IdP) does **not** reproduce the bug — it accepts a wrong `NameID`. Reproduce against a **strict IdP (OneLogin)** and inspect the **content of the `LogoutRequest`**, not just the visual logout.

**Automated**

* `SAMLLogoutRequestContentTest` — `LogoutRequest` `NameID` / `Format` / qualifiers / `SessionIndex`
* `OneLoginSettingsFormatterTest` — provider config → php-saml settings mapping
* `SAMLTest`, `LogoutFromIdpTest`, `LogoutSessionTest` — provider logout ordering, identifier forwarding, IdP callback handling

**Manual (real strict IdP — OneLogin)** — via `centreon-dev-environment` (OneLogin support: centreon/centreon-dev-environment#141, see `doc/centreon-auth.md`):

```
make centreon-saml-install IDP=onelogin
```

1. Log in via the **SAML/SSO** button, then log out.
2. Capture the SP-initiated `LogoutRequest` (the `302` to the IdP SLO carries `?SAMLRequest=`, deflate+base64) and decode `<saml:NameID>`:
   * **Before fix (`develop`):** `NameID = https://.../saml/metadata/... Format="...:entity"`, no `SessionIndex` → rejected by a strict IdP.
   * **After fix (this branch):** `NameID = <user login>` + `<samlp:SessionIndex>` present → SLO accepted, clean logout on both sides.

## Checklist

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

Drago37 committed:
fix logout from saml authentication by adding these keys : samlNameIdFormat, nameIdNameQualifier, samlNameIdSPNameQualifier

Drago37 committed:
fix logging

nfragnet commented on centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php:
```diff
@@ -156,13 +156,22 @@ public function authenticateOrFail(LoginRequest $request): void
         CentreonSession::writeSessionClose('saml', [
             'samlSessionIndex' => $auth->getSessionIndex(),
             'samlNameId' => $auth->getNameId(),
+            'samlNameIdFormat' => ! empty($samlNameIdFormat) ? $samlNameIdFormat : null,
+            'samlNameIdNameQualifier' => ! empty($samlNameIdNameQualifier) ? $samlNameIdNameQualifier : null,
+            'samlNameIdSPNameQualifier' => ! empty($samlNameIdSPNameQualifier) ? $samlNameIdSPNameQualifier : null,
```
Nitpick coding style

```suggestion
            'samlNameIdFormat' => $auth->getNameIdFormat() ?: null,
            'samlNameIdNameQualifier' => $auth->getNameIdNameQualifier() ?: null,
            'samlNameIdSPNameQualifier' => $auth->getNameIdSPNameQualifier() ?: null,
```

nfragnet commented on centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php:
```diff
@@ -156,13 +156,22 @@ public function authenticateOrFail(LoginRequest $request): void
             throw InvalidUserIdAttributeException::create();
         }
 
+        $samlNameIdFormat = $auth->getNameIdFormat();
+        $samlNameIdNameQualifier = $auth->getNameIdNameQualifier();
+        $samlNameIdSPNameQualifier = $auth->getNameIdSPNameQualifier();
+
```
```suggestion
```

Drago37 committed:
Update centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php

Co-authored-by: Nicolas FRAGNET <nfragnet@centreon.com>

Drago37 committed:
Update centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php

Co-authored-by: Nicolas FRAGNET <nfragnet@centreon.com>

Drago37 committed:
api doc fixed

Commit:
fix: change method on logout

Drago37 committed:
fix request to logout

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
invalidate session cache after the logout use case and error handling improved

Drago37 committed:
cs:fix and phpstan:fix

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
unit test fixed

Drago37 committed:
Merge remote-tracking branch 'origin/MON-152191-develop-fix-saml-logout' into MON-152191-develop-fix-saml-logout

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
review of logout by provider and error handling improved

Drago37 committed:
review of logout by provider and error handling improved

Drago37 committed:
Merge remote-tracking branch 'origin/develop' into MON-152191-develop-fix-saml-logout

# Conflicts:
#	centreon/src/Centreon/Domain/Authentication/UseCase/AuthenticateApi.php
#	centreon/src/Core/Security/Authentication/Application/UseCase/Login/LoginResponse.php
#	centreon/src/Core/Security/Authentication/Application/UseCase/LogoutSession/SAML/LogoutFromIdp.php
#	centreon/src/Core/Security/Authentication/Infrastructure/Provider/OpenId.php
#	centreon/src/Core/Security/Authentication/Infrastructure/Provider/ProviderAuthenticationFactory.php
#	centreon/src/Core/Security/Authentication/Infrastructure/Repository/WriteSessionRepository.php
#	centreon/src/Security/WebSSOAuthenticator.php

Drago37 committed:
cs:fix

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Tpo76 commented:
🤖 *This is an automatic comment generated by an internal tool.*

### Complexity Classification: **Standard**

**Justification:** Automatic classification failed; defaulting to standard.

| Metric | Value |
|---|---|
| WIP Points | 2 |
| SLA Track | Standard SLA (2 work days) |
| First review SLA | 1 work day |

A maintainer can override this classification at any time.

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
fix(saml): initiate provider SLO from LogoutSession use case

Move the SAML/OpenID logout initiation out of the SLS callback and into the
LogoutSession use case, triggered before local session invalidation so the SAML
LogoutRequest is built while the SAML identifiers are still in session.

- LogoutSession now initiates SAML logout before invalidate() and OpenID logout
  after it (idToken captured beforehand).
- LogoutFromIdp (/saml/sls) only processes the IdP callback.
- handleCallbackLogoutResponse uses keepLocalSession: false so the local session
  is invalidated on the IdP callback.

Drago37 committed:
test(saml): cover logout flow and make SAML Auth injectable

Introduce SamlAuthFactory to decouple the OneLogin Auth instantiation from the
SAML provider so logout() and handleCallbackLogoutResponse() can be unit tested.

- SamlAuthFactoryInterface/SamlAuthFactory created; SAML uses it instead of new Auth().
- LogoutSessionTest rewritten: SAML logout runs before invalidate, OpenID after,
  local-only and error paths.
- SAMLTest: logout() forwards the SAML identifiers from session;
  handleCallbackLogoutResponse() invalidates the local session (keepLocalSession false).
- LogoutFromIdpTest: SLS callback is processed without re-initiating a logout.

Drago37 committed:
fix(saml): restore findProviderConfiguration on Core ProviderException

The develop merge left ProviderAuthenticationFactory calling
ProviderException::findProviderConfiguration() on the Core exception class,
which only carried unexpectedProvider() after the namespace migration. Re-add
the method so the factory compiles and phpstan passes.

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
test(saml): make LogoutFromIdp unit test runnable under Pest

The @runInSeparateProcess annotation, added to isolate session_start(), crashed
under Pest in CI (PHPUnit\TextUI\Configuration\Registry not found). Move
session_start() from the LogoutFromIdp use case to the SLSController, so the use
case has no native side effect and the test runs without a separate process.

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
test(saml): cover logout request NameID content and settings mapping

Add two unit tests around the MON-152191 SAML logout fix:

- SAMLLogoutRequestContentTest drives the real SAML::logout() through a real
  OneLogin\Saml2\Auth (only redirectTo() neutralised) and asserts the emitted
  <samlp:LogoutRequest> NameID: the user login when session identifiers are
  present, and the IdP entityId + entity format fallback when the session was
  wiped (the original bug).
- OneLoginSettingsFormatterTest pins the config-to-settings mapping, notably
  idp.entityId (the NameID fallback value) and idp.singleLogoutService.url.

a-launois commented on centreon/src/Core/Security/Authentication/Domain/Exception/ProviderException.php:
```diff
@@ -28,12 +28,28 @@ class ProviderException extends \Exception
+     *
+     * @return self
+     */
+    public static function findProviderConfiguration(string $providerConfigurationName, \Throwable $e): self
```
Nitpick: Not a fan of the naming, doesn't sound like "error" to me

Drago37 replied:
fixed

a-launois commented on centreon/doc/API/centreon-api.yaml:
```diff
@@ -273,7 +273,7 @@ paths:
           description: Found
   /authentication/logout:
-    post:
+    get:
```
Why changing this to get ?

Drago37 replied:
The original `POST` was the mistake here. The route was already declared as `GET` (`config/routes/Security/authentication.yaml`); this just aligns the OpenAPI doc with the actual implementation.

`GET` is needed because the SP-initiated SAML SLO relies on a top-level browser redirect (`302` → IdP SLO with `?SAMLRequest=`), which a POST XHR can't follow cross-origin.

Drago37 committed:
refactor(saml): rename ProviderException factory to errorWhileSearchingConfiguration

Drago37 committed:
style(saml): use mb_ltrim in LogoutSessionController for cs compliance

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout

Drago37 committed:
Merge remote-tracking branch 'origin/develop' into MON-152191-develop-fix-saml-logout

# Conflicts:
#	centreon/src/Core/Security/Authentication/Infrastructure/Provider/SAML.php

Drago37 committed:
test(saml): adapt formatter test to RequestStack-based host resolution

The merge brought develop's HttpUrlTrait::getHost(), which now resolves
the host from the injected RequestStack and throws when none is present,
instead of reading $_SERVER. Inject a RequestStack with a Request into
the formatter so the SP entityId resolves, replacing the obsolete
$_SERVER setup.

Drago37 committed:
docs(saml): add authentication providers dev guide

Document the SAML provider in depth (config keys, OneLogin settings
mapping, login/logout flow, the empty-NameID LogoutRequest fallback and
the OneLogin SLO URL subtlety), plus per-provider sections for OpenID
Connect, Web SSO and LDAP.

Drago37 committed:
Merge branch 'develop' into MON-152191-develop-fix-saml-logout
